### PR TITLE
Add architecture documentation

### PR DESCRIPTION
**Pull request by matthewwalsh0**
# Overview

Add architecture documentation including a diagram.

See [here](https://github.com/MetaMask/metamask-desktop/blob/dd40fc1310c4f7de511160198740f810be7c278e/docs/architecture.md) for markdown.

### CODE REVIEW

**Comment by matthewwalsh0** [comment]
I have read the CLA Document and I hereby sign the CLA

File: docs/architecture.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+As the MetaMask extension code is written to be ran in a browser, it often references global objects such as `window`, `document`, and `browserAction`.
+
+In order to avoid changing the MetaMask extension code as much as possible, these global objects are also defined within MetaMask Desktop to perform equivalent operations or to return mock data. See [globals](../packages/app/src/app/globals.ts).
+
+### Node Browser
+
+Some methods still need to be performed in the user's browser, such as displaying a window or changing the extension badge. These method calls are sent to the MetaMask extension to be executed in the real browser, and any return values are sent back to MetaMask Desktop. 
+
+These proxy methods are defined in a whitelist in [node-browser](../packages/app/src/app/browser/node-browser.ts). A second list of unhanled methods is also maintained which simply logs a warning in the console.
```
Comment (gauthierpetetin) [inline docs/architecture.md:27]: unhanled => unhandled

File: docs/architecture.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+### Background Page
+
+In the MetaMask extension, the UI background requests are handled by the background page within the extension.
+
+This is communicated with using the [window.postMessage](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) and [runtime.connect](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/connect) APIs.
+
+### Web Socket Server
+
+All communication with MetaMask desktop is done using a WebSocket server, powered by [ws](https://github.com/websockets/ws).
```
Comment (gauthierpetetin) [inline docs/architecture.md:41]: MetaMask **d**esktop => MetaMask **D**esktop

File: docs/architecture.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+When the background connection is first required, a single WebSocket connection is created from the MetaMask extension to MetaMask Desktop, this is abstracted using the [WebSocketStream](../packages/common/src/web-socket-stream.ts) and the required [encryption](encryption.md) is initialised.
+
+Whenever the extension background page receives a new connection, we immediately pipe the UI stream to the WebSocket stream so all requests received by the background page in the extension are forwarded to MetaMask desktop, and all responses from MetaMask Desktop are ultimately sent back to the extension UI.
```
Comment (gauthierpetetin) [inline docs/architecture.md:45]: MetaMask **d**esktop => MetaMask **D**esktop

File: docs/architecture.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+| Stream | Purpose |
+| --- | --- |
+| browserController | Proxy browser method calls in Node to the browser containing the MetaMask extension. |
+| disable | Request the MetaMask extension stop using MetaMask Desktop, and optionally include the final MetaMask state. |
+| endConnection | Notify MetaMask Desktop when a background connection is no longer required. |
+| newConnection | Notify MetaMask Desktop when a new background connection is required. |
+| pairing | Transfer OTP requests and pairing keys to the extension. |
+| state | Keep the MetaMask state synchronised between the extension and MetaMask Desktop. |
+| version | Verify the MetaMask extension and MetaMask desktop versions are compatible. |
```
Comment (gauthierpetetin) [inline docs/architecture.md:71]: MetaMask **d**esktop => MetaMask **D**esktop

**Review by gauthierpetetin — APPROVED** [review]

File: docs/architecture.md (hunk 2 above)
Comment (matthewwalsh0) [reply docs/architecture.md]: Done.

File: docs/architecture.md (hunk 3 above)
Comment (matthewwalsh0) [reply docs/architecture.md]: Done.

File: docs/architecture.md (hunk 4 above)
Comment (matthewwalsh0) [reply docs/architecture.md]: Done.

**Review by vinistevam — APPROVED** [review]

**Pull request merged**